docs/source/debugging.rst

```diff
@@ -0,0 +1,237 @@
+   x.compute(rerun_exceptions_locally=True)
+
+On the distributed scheduler use the ``recreate_error_locally`` method on
+anything that contains ``Futures`` :
+
+.. code-block:: python
+
+   >>> x.compute()
+   ZeroDivisionError(...)
+
+   >>> %pdb
+   >>> future = client.compute(x)
+   >>> client.recreate_error_locally(future)
+
+
+Remove Failed Futures Manually
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Sometimes only parts of your computations fail, for example if some rows of a
+CSV dataset are faulty in some way.  When running with the distributed
+scheduler you can remove chunks of your data that have produced bad results if
+you switch to dealing with Futures.
+
+.. code-block:: python
+
+   >>> import dask.dataframe as dd
+   >>> df = ...           # create dataframe
+   >>> df = df.persist()  # start computing on the cluster
+
+   >>> from distributed.client import futures_of
+   >>> futures = futures_of(df)  # get futures behind dataframe
+   >>> futures
+   [<Future: status: finished, type: pd.DataFrame, key: load-1>
+    <Future: status: finished, type: pd.DataFrame, key: load-2>
+    <Future: status: error, key: load-3>
+    <Future: status: pending, key: load-4>
+    <Future: status: error, key: load-5>]
+
+   >>> # wait until computation is done
+   >>> while any(f.status == 'pending' for f in futures):
+   ...     sleep(0.1)
+
+   >>> # pick out only the successful futures and reconstruct the dataframe
+   >>> good_futures = [f for f in futures if f.status == 'finished']
+   >>> df = dd.from_delayed(good_futures, meta=df._meta)
+
+This is a bit of a hack, but often practical when first exploring messy data.
+If you are using the concurrent.futures API (map, submit, gather) then this
+approach is more natural.
+
+
+Inspect Scheduling State
+------------------------
+
+Not all errors present themselves as Exceptions.  For example in a distributed
+systems workers may die unexpectedly or your computation may be unreasonably
+slow due to inter-worker communication or scheduler overhead or one of several
+other issues.  Getting feedback about what's going on can help to identify
+both failures and general performance bottlenecks.
+
+For the single-machine scheduler see :doc:`diagnostics <diagnostics`
+documentation.  The rest of the section will assume that you are using the
+`distributed scheduler <http://distributed.readthedocs.io/en/latest/>`_ where
+these issues arise more commonly.
+
+Web Diagnostics
+~~~~~~~~~~~~~~~
+
+First, the distributed scheduler has a number of `diagnostic web pages
+<http://distributed.readthedocs.io/en/latest/web.html>`_ showing dozens of
+recorded metrics like CPU, memory, network, and disk use, a history of previous
+tasks, allocation of tasks to workers, worker memory pressure, work stealing,
+open file handle limits, etc..  *Many* problems can be correctly diagnosed by
+inspecting these pages.  By default these are available at
+``http://scheduler:8787/`` ``http://scheduler:8788`` and ``http://worker:8789``
```

Comment:
Worth saying what's at these default ports ? Also the 8787 has a trailing slash, the others don't.